compiler/src/dotty/tools/dotc/transform/CoverageTransformMacro.scala

```diff
@@ -0,0 +1,204 @@
+import dotty.tools.dotc.coverage.Coverage
+import dotty.tools.dotc.coverage.Statement
+import dotty.tools.dotc.coverage.Serializer
+import dotty.tools.dotc.coverage.Location
+import dotty.tools.dotc.core.Symbols.defn
+import dotty.tools.dotc.core.Symbols.Symbol
+import dotty.tools.dotc.core.Decorators.toTermName
+import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.typer.LiftCoverage
+
+import scala.quoted
+
+/** Phase that implements code coverage, executed when the "-coverage
+ *  OUTPUT_PATH" is added to the compilation.
+ */
+class CoverageTransformMacro extends MacroTransform with IdentityDenotTransformer {
+  import ast.tpd._
+
+  override def phaseName = "coverage"
+
+  // Atomic counter used for assignation of IDs to difference statements
+  val statementId = new AtomicInteger(0)
+
+  var outputPath = ""
+
+  // Main class used to store all instrumented statements
+  val coverage = new Coverage
+
+  override def run(using ctx: Context): Unit = {
+
+    if (ctx.settings.coverageOutputDir.value.nonEmpty) {
+      outputPath = ctx.settings.coverageOutputDir.value
+
+      // Ensure the dir exists
+      val dataDir = new File(outputPath)
+      val newlyCreated = dataDir.mkdirs()
+
+      if (!newlyCreated) {
+        // If the directory existed before, let's clean it up.
+        dataDir.listFiles
+          .filter(_.getName.startsWith("scoverage"))
+          .foreach(_.delete)
+      }
+
+      super.run
+
+
+      Serializer.serialize(coverage, outputPath, ctx.settings.coverageSourceroot.value)
+    }
+  }
+
+  protected def newTransformer(using Context): Transformer =
+    new CoverageTransormer
+
+  class CoverageTransormer extends Transformer {
+    var instrumented = false
+
+    override def transform(tree: Tree)(using Context): Tree = {
+      tree match {
+        case tree: If =>
+          cpy.If(tree)(
+            cond = transform(tree.cond),
+            thenp = instrument(transform(tree.thenp), branch = true),
+            elsep = instrument(transform(tree.elsep), branch = true)
+          )
+        case tree: Try =>
+          cpy.Try(tree)(
+            expr = instrument(transform(tree.expr), branch = true),
+            cases = instrumentCasees(tree.cases),
+            finalizer = instrument(transform(tree.finalizer), true)
+          )
+        case Apply(fun, _)
+            if (
+              fun.symbol.exists &&
+                fun.symbol.isInstanceOf[Symbol] &&
+                fun.symbol == defn.Boolean_&& || fun.symbol == defn.Boolean_||
+            ) =>
+          super.transform(tree)
+        case tree @ Apply(fun, args) if (fun.isInstanceOf[Apply]) =>
+          // We have nested apply, we have to lift all arguments
+          // Example: def T(x:Int)(y:Int)
+          // T(f())(1) // should not be changed to {val $x = f(); T($x)}(1) but to {val $x = f(); val $y = 1; T($x)($y)}
```

Comment:
I have the same question, actually. I believe that it would be simpler to insert Invoker calls "in place".
@ckipp01 was there a specific reason to lift the arguments instead of replacing, for example, `f(x)` with `f(Invoker.invoked(someParams, x))`?

Reply:
I think the tricky thing here is dealing with the instrumentation of f itself, if you replace:
```scala
f(a(), b())
```
by:
```scala
{invoked(1); f}({invoked(2); a()}, {invoked(3); b()}}
```
Then you will consider `f` as instrumented even if executing `a()` or `b()` throws an exception and prevents `f` from actually being run at all, on the other hand if you delay the `invoked` call until after the actual method call, you won't record the method as having been called if it threw an exception itself. See Figures 4 and 5 of http://guillaume.martres.me/code_coverage.pdf for an example.

In practice, we might decide that the complexity required to support this is too high and we're willing to accept some imprecision in our coverage reports, or perhaps there is some alternative strategy we could use (could we have the backend take care of inserting the `invoked` call for method calls? [`genCallMethod`](https://github.com/lampepfl/dotty/blob/1b25f653855ac485f9e0513be9b27c0f29725f6f/compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala#L1300) is called after we've already pushed the method arguments on the stack)

Reply:
Hmm, one thing to be careful of here is inline def handling. Inlined nodes are currently dropped in Erasure: https://github.com/lampepfl/dotty/blob/1b25f653855ac485f9e0513be9b27c0f29725f6f/compiler/src/dotty/tools/dotc/transform/Erasure.scala#L875-L878 so any phase after that cannot emit coverage information for inline defs.

Reply:
how about
```scala
try
  f({invoked(2); a()}, ...)
finally
  invoked(1)
```
?

Reply:
What do you think about modifying the definition of f to take care of `invoked`?
```
def f(arg1: T1, arg2: T2) =
  invoked(id)
  originalBodyOfF, which uses arg1, arg2
```
We can't modify external libraries but we don't try to measure their coverage so it's not a big problem.

Reply:
I don't see how that would work, if I'm calling `f(a)` I want to know if this particular call to `f` was ever reached to color it green (see figure 12 in http://guillaume.martres.me/code_coverage.pdf)

Reply:
Using a try/finally might be worth a try indeed.

Reply:
I would rather not do it in the backend, that would force the work to be redone for Scala.JS and Scala Native

Reply:
> I don't see how that would work, if I'm calling `f(a)` I want to know if this particular call to `f` was ever reached to color it green (see figure 12 in http://guillaume.martres.me/code_coverage.pdf)

The new `f` could take an argument that indicate its calling location.
It would work for methods that are called by reflection too, e.g. serialization frameworks, which is not covered by the current Invoker. In the case of reflection the calling location would be unknown, of course. However, that would add a parameter and I wonder if it's a good idea (with regards to the parameter limit, for instance).

Reply:
Something that would work is to let the inliner perform the lifting instead of doing it ourselves for the coverage.
Given a code like this:
```scala
f(a(), b())
```
Emit this:
```scala
inline def f$instrumented(a: A, b: B): R =
  Invoker.invoke(id, outputDir)
  f(a, b)
f$instrumented(a(), b())
```
This would remove the need for `LiftCoverage` and ensure that all the corner cases are properly handled.

But I don't know how to generate such a method with the paramss of `f`. Here we have a `tree: Apply`. I've tried `tree.fun.symbol.defTree`, hoping to get the `DefDef` of the method, but it's always empty.

Reply:
I think generating defs is likely to be a lot more complicated than alternative solutions